# Making Request to Patient Resource

In order to make a request from using Python, we leverage the `requests` library. To form the request, we need three things: base URL, resource name, and parameters.

## Components of the Request

### 1. Base URL
The base URL is the address of the FHIR server where the resources are hosted. This is the starting point for any FHIR request and directs the request to the correct server.

### 2. Resource Name
The resource name specifies the type of FHIR resource you want to interact with. FHIR defines many resource types such as `Patient`, `Observation`, `Medication`, etc. The resource name is appended to the base URL to form the endpoint for the request.

### 3. Parameters
Parameters are used to refine and control the data returned by the FHIR server. They can specify various criteria such as format, filters, and other query options.

Common parameters include:
- `_format`: Specifies the response format (e.g., `json`, `xml`).
- `_pretty`: Formats the response for readability.

Every API endpoint has different parameters. Usually they are outlined in the documentation. For the `Patient` Resource one such parameter is `birthday`. This means we can query for patient data by birthday

In [1]:
import requests

# Base URL
base_url = 'http://fhirserver.hl7fundamentals.org/fhir'

# Resource name
resource = 'Patient'

# Parameters for the request
params = {
    'birthdate': 'ge1980-01-01',  # Example birthdate
    '_format': 'json',
    '_pretty': 'true'
}

# Construct the request URL
patient_request_url = f"{base_url}/{resource}"

# Print the request URL for debugging purposes
print('Requesting Patient Resource from:', patient_request_url)

# Perform the request with parameters
response = requests.get(patient_request_url, params=params)

# Check if the request was successful
if response.status_code == 200:
    # Print the response in JSON format
    patient_data = response.json()
else:
    print(f"Failed to retrieve data: {response.status_code}")

# Example: Extract and print specific patient information
if 'entry' in patient_data:
    for entry in patient_data['entry']:
        patient = entry['resource']
        print(f"Patient ID: {patient.get('id', 'N/A')}")
        print(f"Name: {patient.get('name', 'N/A')}")
        print(f"Gender: {patient.get('gender', 'N/A')}")
        print(f"Birthdate: {patient.get('birthDate', 'N/A')}")
else:
    print("No patient data found.")


Requesting Patient Resource from: http://fhirserver.hl7fundamentals.org/fhir/Patient
Patient ID: 3282
Name: [{'use': 'usual', 'text': 'Brian Ahern', 'family': 'Ahern', 'given': ['Brian'], 'prefix': ['Mr']}, {'use': 'nickname', 'given': ['Tall Man']}]
Gender: male
Birthdate: 1980-01-01
Patient ID: 4117
Name: [{'use': 'official', 'text': 'Don Taylor', 'family': 'Taylor', 'given': ['Don']}]
Gender: male
Birthdate: 1980-01-01
Patient ID: 11774
Name: [{'use': 'official', 'text': 'Juana Melendi', 'family': 'Melendi', 'given': ['Juana']}]
Gender: female
Birthdate: 1980-01-01
Patient ID: 20182
Name: [{'use': 'official', 'family': 'Tsai', 'given': ['James']}]
Gender: male
Birthdate: 1980-01-01
Patient ID: 123452
Name: [{'family': 'Villahermosa', 'given': ['Gabriel']}]
Gender: male
Birthdate: 1980-01-01
Patient ID: 124250
Name: [{'family': 'Zhang', 'given': ['Jinhua']}]
Gender: female
Birthdate: 1980-01-01
Patient ID: 124252
Name: [{'family': 'Zhang', 'given': ['Jinhua']}]
Gender: female
Birthda

# Activity

Please look for another parameter for the Patient resource to filter by and create a new request to fetch patient data.

## Step-by-Step Instructions

1. **Research Additional Parameters**:
   - Visit [FHIR Patient Resource](http://fhirserver.hl7fundamentals.org/resource?serverId=home&pretty=true&_summary=&resource=Patient) to find other parameters that can be used to filter patient data.

2. **Choose a Parameter**:
   - Select one parameter from the documentation. For this example, let's choose `gender` to filter patients by gender.

3. **Formulate the Request**:
   - Update the `params` dictionary to include the new parameter along with the existing ones.


In [2]:
import requests

# Base URL
base_url = 'http://fhirserver.hl7fundamentals.org/fhir'

# Resource name
resource = 'Patient'

# Parameters for the request
params = {
    'gender': 'female',
    '_format': 'json',
    '_pretty': 'true'
}

# Construct the request URL
patient_request_url = f"{base_url}/{resource}"

# Print the request URL for debugging purposes
print('Requesting Patient Resource from:', patient_request_url)

# Perform the request with parameters
response = requests.get(patient_request_url, params=params)

# Check if the request was successful
if response.status_code == 200:
    # Print the response in JSON format
    patient_data = response.json()
else:
    print(f"Failed to retrieve data: {response.status_code}")

# Example: Extract and print specific patient information
if 'entry' in patient_data:
    for entry in patient_data['entry']:
        patient = entry['resource']
        print(f"Patient ID: {patient.get('id', 'N/A')}")
        print(f"Name: {patient.get('name', 'N/A')}")
        print(f"Gender: {patient.get('gender', 'N/A')}")
        print(f"Birthdate: {patient.get('birthDate', 'N/A')}")
else:
    print("No patient data found.")


Requesting Patient Resource from: http://fhirserver.hl7fundamentals.org/fhir/Patient
Patient ID: X12984
Name: [{'use': 'official', 'text': 'Preut Michaela', 'family': 'Gonzalez Serrano', 'given': ['Preut Michaela Testing']}, {'use': 'nickname', 'given': ['Mick']}]
Gender: female
Birthdate: 1920-12-05
Patient ID: 75
Name: [{'family': 'Al Hamad', 'given': ['Mona', 'Mohammed']}]
Gender: female
Birthdate: 2022-12-31
Patient ID: 93
Name: [{'family': 'Al Hamad', 'given': ['Mona', 'Mohammed']}]
Gender: female
Birthdate: 2022-12-31
Patient ID: 94
Name: [{'family': 'Al Hamad', 'given': ['Mona', 'Mohammed']}]
Gender: female
Birthdate: 2022-12-31
Patient ID: 95
Name: [{'family': 'Al Hamad', 'given': ['Mona', 'Mohammed']}]
Gender: female
Birthdate: 2022-12-31
Patient ID: 97
Name: [{'family': 'Al Hamad', 'given': ['Mona', 'Mohammed']}]
Gender: female
Birthdate: 2022-12-31
Patient ID: 112
Name: [{'use': 'official', 'text': 'Juan Manuel Ricardo', 'family': 'Ricardo', 'given': ['Juan Manuel']}]
Gender

In [3]:
response.json()

{'resourceType': 'Bundle',
 'id': '47e95fcc-6bd9-4569-b067-46759fb6b94b',
 'meta': {'lastUpdated': '2024-07-11T14:05:19.677+00:00'},
 'type': 'searchset',
 'link': [{'relation': 'self',
   'url': 'http://fhirserver.hl7fundamentals.org/fhir/Patient?_format=json&_pretty=true&gender=female'},
  {'relation': 'next',
   'url': 'http://fhirserver.hl7fundamentals.org/fhir?_getpages=47e95fcc-6bd9-4569-b067-46759fb6b94b&_getpagesoffset=20&_count=20&_format=json&_pretty=true&_bundletype=searchset'}],
 'entry': [{'fullUrl': 'http://fhirserver.hl7fundamentals.org/fhir/Patient/X12984',
   'resource': {'resourceType': 'Patient',
    'id': 'X12984',
    'meta': {'versionId': '11',
     'lastUpdated': '2024-06-16T08:27:50.951+00:00',
     'source': '#rXwHzsqOFnEEV8gG'},
    'text': {'status': 'generated',
     'div': '<div xmlns="http://www.w3.org/1999/xhtml"><p>Patient 12984</p></div>'},
    'identifier': [{'use': 'usual',
      'system': 'https://www.national-office.gov/ni',
      'value': '77000000

# Medical Request

Create a simple medical request for a patient using Epic on FHIR. Follow the steps below to complete the task.

1. **Select a Patient**:
    - Use the Epic FHIR Sandbox to select a patient. You can do this by searching for patients or using a specific patient ID provided in the sandbox environment.
    - Note down the patient ID for use in your request.

2. **Identify the Base URL Endpoint**:
    - The base URL for making FHIR requests in the Epic Sandbox environment is typically: `http://fhirserver.hl7fundamentals.org/fhir"`
    - Ensure you have the correct endpoint for your requests.

3. **Create a Medical Request**:
    - Use the patient ID and the base URL to create a medical request. Below is an example of how to structure your request in Python using the `requests` library.

In [4]:
import requests

# Step 1: Select a Patient
patient_id = '6636'  # Example patient ID

# Step 2: Identify the Base URL Endpoint
base_url = 'http://fhirserver.hl7fundamentals.org/fhir'

# Step 3: Create a Medical Request
# Resource name
resource = 'MedicationRequest'

# Parameters for the request
params = {
    'patient': patient_id,  # Patient ID
    '_format': 'json',
}

# Construct the request URL
patient_request_url = f"{base_url}/{resource}"

# Print the request URL for debugging purposes
print('Requesting Patient Resource from:', patient_request_url)

# Perform the request with parameters
response = requests.get(patient_request_url, params=params)

# Check if the request was successful
if response.status_code == 200:
    # Print the response in JSON format
    patient_data = response.json()
    print(patient_data)
else:
    print(f"Failed to retrieve data: {response.status_code}")

Requesting Patient Resource from: http://fhirserver.hl7fundamentals.org/fhir/MedicationRequest
{'resourceType': 'Bundle', 'id': '2c557f10-200d-417c-a1e2-19ff17ff1e1b', 'meta': {'lastUpdated': '2024-07-11T14:05:21.329+00:00'}, 'type': 'searchset', 'total': 1, 'link': [{'relation': 'self', 'url': 'http://fhirserver.hl7fundamentals.org/fhir/MedicationRequest?_format=json&patient=6636'}], 'entry': [{'fullUrl': 'http://fhirserver.hl7fundamentals.org/fhir/MedicationRequest/6714', 'resource': {'resourceType': 'MedicationRequest', 'id': '6714', 'meta': {'versionId': '1', 'lastUpdated': '2023-11-14T15:02:06.216+00:00', 'source': '#CoyI88SBMDWO7lVE'}, 'status': 'active', 'intent': 'order', 'medicationCodeableConcept': {'coding': [{'system': 'http://snomed.info/sct', 'code': '386910003', 'display': 'Anastrozol'}]}, 'subject': {'reference': 'Patient/6636'}, 'authoredOn': '2022-04-01T10:12:14', 'dosageInstruction': [{'timing': {'repeat': {'frequency': 1, 'period': 84, 'periodUnit': 'd'}}, 'route': 

In [5]:
response.json()

{'resourceType': 'Bundle',
 'id': '2c557f10-200d-417c-a1e2-19ff17ff1e1b',
 'meta': {'lastUpdated': '2024-07-11T14:05:21.329+00:00'},
 'type': 'searchset',
 'total': 1,
 'link': [{'relation': 'self',
   'url': 'http://fhirserver.hl7fundamentals.org/fhir/MedicationRequest?_format=json&patient=6636'}],
 'entry': [{'fullUrl': 'http://fhirserver.hl7fundamentals.org/fhir/MedicationRequest/6714',
   'resource': {'resourceType': 'MedicationRequest',
    'id': '6714',
    'meta': {'versionId': '1',
     'lastUpdated': '2023-11-14T15:02:06.216+00:00',
     'source': '#CoyI88SBMDWO7lVE'},
    'status': 'active',
    'intent': 'order',
    'medicationCodeableConcept': {'coding': [{'system': 'http://snomed.info/sct',
       'code': '386910003',
       'display': 'Anastrozol'}]},
    'subject': {'reference': 'Patient/6636'},
    'authoredOn': '2022-04-01T10:12:14',
    'dosageInstruction': [{'timing': {'repeat': {'frequency': 1,
        'period': 84,
        'periodUnit': 'd'}},
      'route': {'cod

# Vital Signs

Create a Python script to retrieve only the patient's numeric vital signs without components, sorted by date and code. Include the result value rounded to 2 decimals and other mandatory elements.

1) The order of the elements should be:
   - `effectiveDateTime`
   - `code.text`
   - `valueQuantity.value` and `valueQuantity.unit` (formatted as 'value unit')
2) Do not include patient or category information, as they are used as search parameters.
3) For numeric data, include only the value and the unit separated by a space (e.g., '18 g/l').
4) For CodeableConcept elements, include only the text element.
5) Use patient id `X12984`

Example: Each line represents a different Vital Sign:

mandatory_element_1|...|mandatory_element_n|...
mandatory_element_2|...|mandatory_element_n|...

In [10]:
import requests

# Patient id
id = 'X12984'

# Base URL
base_url = 'http://fhirserver.hl7fundamentals.org/fhir'
resource = 'Observation'

# Parameters for the request
params = {
    "patient": id,
    "_format": "json",
}

# Construct the request URL
observation_request = f"{base_url}/{resource}"

# Print the request URL for debugging purposes
print('Now Searching for Observation...@', observation_request)

# Perform the request with parameters
response = requests.get(observation_request, params=params)
procedure_response = response.json()

# Initialize the vital_signs list
vital_signs = []

for entry in procedure_response['entry']:
    resource = entry['resource']
    effectiveDateTime = resource.get('effectiveDateTime')
    
    valueQuantity = resource['valueQuantity']
    value = valueQuantity['value']
    unit = valueQuantity['unit']
    
    code = resource['code']['coding'][0]
    display = code['display']
    
    vital_signs.append(f'{effectiveDateTime}|{display}|{value} {unit}|')
    
    
    
for vital_sign in vital_signs:
    print(vital_sign)




Now Searching for Observation...@ http://fhirserver.hl7fundamentals.org/fhir/Observation
2013-10-14T14:15:16-04:00|Pain severity - 0-10 verbal numeric rating [Score] - Reported|0.57 {score}|
2004-10-18T14:15:16-04:00|Body Height|173.5 cm|
2004-10-18T14:15:16-04:00|Pain severity - 0-10 verbal numeric rating [Score] - Reported|2.81 {score}|
2013-10-14T14:15:16-04:00|Body Weight|91.76 kg|
2010-11-01T14:15:16-04:00|Pain severity - 0-10 verbal numeric rating [Score] - Reported|3.77 {score}|
2012-10-08T14:15:16-04:00|Body Height|173.5 cm|
2010-11-01T14:15:16-04:00|Body Weight|91.76 kg|
2010-11-01T14:15:16-04:00|Body Height|173.5 cm|
2012-10-08T14:15:16-04:00|Body Mass Index|30.48 kg/m2|
2004-10-18T14:15:16-04:00|Body Weight|88.01 kg|
2006-11-27T13:15:16-05:00|Body Mass Index|30.48 kg/m2|
2006-10-23T14:15:16-04:00|Body Mass Index|30.48 kg/m2|
2008-10-27T14:15:16-04:00|Body Weight|91.76 kg|
2008-10-27T14:15:16-04:00|Pain severity - 0-10 verbal numeric rating [Score] - Reported|2.92 {score}|
20

In [9]:
response.json()['entry'][0]

{'fullUrl': 'http://fhirserver.hl7fundamentals.org/fhir/Observation/47',
 'resource': {'resourceType': 'Observation',
  'id': '47',
  'meta': {'versionId': '1',
   'lastUpdated': '2023-10-17T14:55:47.837+00:00',
   'source': '#Jk5LnLpSEi548SYI'},
  'status': 'final',
  'category': [{'coding': [{'system': 'http://terminology.hl7.org/CodeSystem/observation-category',
      'code': 'vital-signs',
      'display': 'Vital Signs'}],
    'text': 'Vital Signs'}],
  'code': {'coding': [{'system': 'http://loinc.org',
     'code': '72514-3',
     'display': 'Pain severity - 0-10 verbal numeric rating [Score] - Reported'}],
   'text': 'Pain severity - 0-10 verbal numeric rating [Score] - Reported'},
  'subject': {'reference': 'Patient/X12984'},
  'effectiveDateTime': '2013-10-14T14:15:16-04:00',
  'issued': '2013-10-14T14:15:16-04:00',
  'valueQuantity': {'value': 0.57,
   'unit': '{score}',
   'system': 'http://unitsofmeasure.org',
   'code': '{score}'}},
 'search': {'mode': 'match'}}

# Clinical Goals

Create a Python script to retrieve the patient's clinical goals, including all mandatory elements except the 'subject' element, and return them separated by | (pipe).

1) The order of the elements should be:
   - `startDate`
   - `description`
   - `dueDate`
   - `lifecycleStatus`
2) Do not include patient or category information, as they are used as search parameters.
3) For string data, include the text directly.
4) For CodeableConcept elements, include only the text element.
5) Use patient id `12984`.

Example: Each line represents a different Goal:

mandatory_element_1|...|mandatory_element_n|...
mandatory_element_2|...|mandatory_element_n|...

In [ ]:
import requests

# Patient id
id = 'X12984'

# Base URL
base_url = 'http://fhirserver.hl7fundamentals.org/fhir'
resource = 'Goal'

# Parameters for the request
params = {
    "patient": id,
    "_format": "json"
}

# Construct the request URL
goal_request = f"{base_url}/{resource}"

# Print the request URL for debugging purposes
print('Now Searching for Goals...@', goal_request)

# Perform the request with parameters
response = requests.get(goal_request, params=params)
goal_response = response.json()

# Initialize the goals list
goals = []

# Extracting and formatting the goals
for entry in goal_response['entry']:
    resource = entry['resource']
    start_date = resource.get('startDate', '')
    description = resource.get('description', {}).get('text', '')
    
    # Initialize target values
    target_value = ''
    target_unit = ''
    
    # Check if target is a list
    if 'target' in resource and isinstance(resource['target'], list):
        for target in resource['target']:
            if 'dueDate' in target:
                dueDate = target.get('dueDate')
    
    status = resource.get('lifecycleStatus', '')
    
    formatted_goal = f"{start_date}|{description}|{dueDate}|{status}"
    goals.append(formatted_goal)

# Output the formatted goals
for goal in goals:
    print(goal)


In [ ]:
target.get('dueDate')

# Implantable Devices

Create a Python script to retrieve the mandatory/must-support elements for all US CORE FHIR R4 Implantable Devices for a patient. Include the elements in the order they appear in the Device resource definition, separated by | (pipe).

1) The order of the elements should be:
   - `deviceIdentifier`
   - `expirationDate`
   - `lotNumber`
   - `modelNumber`
   - `device_type`
2) For CodeableConcept elements, include only the display element.
3) Use patient id `1`.

Example: Each line represents a different Device resource:

mandatory_ms_element_1|...|mandatory_ms_element_n|...
mandatory_ms_element_2|...|mandatory_ms_element_n|...

In [ ]:
import requests

# Patient id
id = '1'

# Base URL
base_url = 'http://fhirserver.hl7fundamentals.org/fhir'
resource = 'Device'

# Parameters for the request
params = {
    "patient": id,
    "_format": "json"
}

# Construct the request URL
device_request = f"{base_url}/{resource}"

# Print the request URL for debugging purposes
print('Now Searching for Devices...@', device_request)

# Perform the request with parameters
response = requests.get(device_request, params=params)
device_response = response.json()

# Initialize the devices list
devices = []

# Extracting and formatting the devices
for entry in device_response['entry']:
    resource = entry['resource']
    identifier = resource.get('identifier', [])
    system = None
    if identifier:
        system = identifier[0].get('system')
    udi_carrier = resource.get('udiCarrier', [{}])[0] if resource.get('udiCarrier') else {}
    expiration_date = resource.get('expirationDate', '')
    lot_number = resource.get('lotNumber', '')
    serial_number = resource.get('modelNumber', '')
    device_type = resource.get('type', {}).get('coding', [{}])[0].get('display', '')

    formatted_device = f"{system}|{expiration_date}|{lot_number}|{serial_number}|{device_type}"
    devices.append(formatted_device)

# Output the formatted devices
for device in devices:
    print(device)
